src/main/java/com/cleanroommc/groovyscript/compat/mods/advancedrocketry/BaseRegistry.java

```diff
@@ -0,0 +1,230 @@
+            return removeByFluidInput((FluidStack) outputItem);
+        }
+        return getRecipeList().removeIf(r -> {
+            List<ItemStack> output = r.getOutput();
+            if (output.stream().anyMatch(outputItem)) {
+                addBackup(r);
+                return true;
+            }
+            return false;
+        });
+    }
+
+    @MethodDescription(priority = 2000, example = @Example(commented = true))
+    public void removeAll() {
+        List<IRecipe> recipes = getRecipeList();
+        recipes.forEach(this::addBackup);
+        recipes.clear();
+    }
+
+    @MethodDescription(type = MethodDescription.Type.QUERY)
+    public SimpleObjectStream<IRecipe> streamRecipes() {
+        List<IRecipe> recipes = getRecipeList();
+        return new SimpleObjectStream<>(recipes)
+                .setRemover(this::removeRecipe);
+    }
+
+    public abstract static class RecipeBuilder extends AbstractRecipeBuilder<IRecipe> {
+        // still have to override the Validate method + add getRegistry method
+
+        @Property(value = "groovyscript.wiki.advancedrocketry.power.value", needsOverride = true)
+        protected int power = 0;
+        @Property(value = "groovyscript.wiki.advancedrocketry.time.value", needsOverride = true)
+        protected int time = 0;
+        @Property(value = "groovyscript.wiki.advancedrocketry.outputSize.value", needsOverride = true)
+        protected int outputSize = 0;
+
+        protected final List<Float> outputChances = new ArrayList<>();
+
+        // 0.0f is used because that's what AR uses for "100% chance". Copium.
+        @Override
+        @RecipeBuilderMethodDescription
+        public RecipeBuilder output(ItemStack output) {
+            return output(output, 0.0f);
+        }
+
+        @RecipeBuilderMethodDescription(field = {"output", "outputChances"})
+        protected RecipeBuilder output(ItemStack output, float chance) {
+            this.output.add(output);
+            this.outputChances.add(chance);
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return String.format("Error adding Advanced Rocketry %s recipe", getRegistry().getName());
+        }
+
+        protected int getHatchesNeeded() {
+            return (int) (Math.ceil(((double) input.size()) / 4) + Math.ceil(((double) output.size()) / 4) + fluidInput.size() + fluidOutput.size());
+        }
+
+        protected abstract BaseRegistry getRegistry();
+
+        @Override
+        @RecipeBuilderRegistrationMethod
+        public @Nullable IRecipe register() {
+            if (!validate()) return null;
+            List<List<ItemStack>> inputs = new LinkedList<>();
+            Map<Integer, String> oredicts = new HashMap<>();
+            for (int i = 0; i < input.size(); i++) {
+                IIngredient in = input.get(i);
+                inputs.add(Arrays.stream(in.getMatchingStacks()).collect(Collectors.toList()));

```

Comment:
`Arrays.asList()`?

Reply:
gonna look at these `collectors.toList()` uses and fix

Reply:
done